server/events_test.go

```diff
@@ -1209,3 +1210,89 @@ func TestGatewayNameClientInfo(t *testing.T) {
 		t.Fatalf("Expected a cluster name of 'TEST CLUSTER 22', got %q", info.Cluster)
 	}
 }
+
+type slowAccResolver struct {
+	sync.Mutex
+	AccountResolver
+	acc string
+}
+
+func (sr *slowAccResolver) Fetch(name string) (string, error) {
+	sr.Lock()
+	delay := sr.acc == name
+	sr.Unlock()
+	if delay {
+		time.Sleep(200 * time.Millisecond)
+	}
+	return sr.AccountResolver.Fetch(name)
+}
+
+func TestFetchAccountRace(t *testing.T) {
+	sa, oa, sb, ob, _ := runTrustedGateways(t)
+	defer sa.Shutdown()
+	defer sb.Shutdown()
+
+	// Let's create a user account.
+	okp, _ := nkeys.FromSeed(oSeed)
+	akp, _ := nkeys.CreateAccount()
+	pub, _ := akp.PublicKey()
+	nac := jwt.NewAccountClaims(pub)
+	jwt, _ := nac.Encode(okp)
+	userAcc := pub
+
+	// Replace B's account resolver with one that introduces
+	// delay during the Fetch()
+	sb.mu.Lock()
+	sac := &slowAccResolver{AccountResolver: sb.accResolver}
+	sb.accResolver = sac
+	sb.mu.Unlock()
+
+	// Add the account in sa and sb
+	addAccountToMemResolver(sa, userAcc, jwt)
+	addAccountToMemResolver(sb, userAcc, jwt)
+
+	// Tell the slow account resolver which account to slow down
+	sac.Lock()
+	sac.acc = userAcc
+	sac.Unlock()
+
+	urlA := fmt.Sprintf("nats://%s:%d", oa.Host, oa.Port)
+	urlB := fmt.Sprintf("nats://%s:%d", ob.Host, ob.Port)
+
+	nca, err := nats.Connect(urlA, createUserCreds(t, sa, akp))
+	if err != nil {
+		t.Fatalf("Error connecting to A: %v", err)
+	}
+	defer nca.Close()
+
+	// Since there is optimistic send, this message will go to B
+	// and on processing this message, B will lookup/fetch this
+	// account.
```

Comment:
Should we say also that the system account will send a notification about the account to B as well?

Reply:
Updated